src/test/java/seedu/address/logic/commands/CompleteCommandTest.java

```diff
@@ -0,0 +1,193 @@
+        expectedModel.undoTaskManager();
+        assertCommandSuccess(new UndoCommand(), model, commandHistory, UndoCommand.MESSAGE_SUCCESS, expectedModel);
+
+        // redo -> same first person deleted again
+        expectedModel.redoTaskManager();
+        assertCommandSuccess(new RedoCommand(), model, commandHistory, RedoCommand.MESSAGE_SUCCESS, expectedModel);
+    }
+
+    @Test
+    public void executeUndoRedo_invalidIndexUnfilteredList_failure() {
+        Index outOfBoundIndex = Index.fromOneBased(model.getFilteredTaskList().size() + 1);
+        CompleteCommand completeCommand = new CompleteCommand(outOfBoundIndex);
+
+        // execution failed -> task manager state not added into model
+        assertCommandFailure(completeCommand, model, commandHistory, Messages.MESSAGE_INVALID_TASK_DISPLAYED_INDEX);
+
+        // single task manager state in model -> undoCommand and redoCommand fail
+        assertCommandFailure(new UndoCommand(), model, commandHistory, UndoCommand.MESSAGE_FAILURE);
+        assertCommandFailure(new RedoCommand(), model, commandHistory, RedoCommand.MESSAGE_FAILURE);
+    }
+
+    /**
+     * 1. Deletes a {@code Task} from a filtered list.
+     * 2. Undo the deletion.
+     * 3. The unfiltered list should be shown now. Verify that the index of the previously
+     * deleted task in the unfiltered list is different from the index at the filtered list.
+     * 4. Redo the deletion. This ensures {@code RedoCommand} deletes the task object regardless of
+     * indexing.
+     */
+    @Test
+    public void executeUndoRedo_validIndexFilteredList_sameTaskCompleted() throws Exception {
+        CompleteCommand completeCommand = new CompleteCommand(INDEX_FIRST_TASK);
+        Model expectedModel = new ModelManager(model.getTaskManager(), new UserPrefs());
+
+        showTaskAtIndex(model, INDEX_SECOND_TASK);
+        //showTaskAtIndex(expectedModel, INDEX_SECOND_TASK);
+        Task taskToComplete = model.getFilteredTaskList().get(INDEX_FIRST_TASK.getZeroBased());
+        expectedModel.updateTask(taskToComplete, simpleCompleteTask(taskToComplete));
+        expectedModel.commitTaskManager();
+
+        // completes -> completes second task in unfiltered task list / first task in filtered task
+        // list
+        completeCommand.execute(model, commandHistory);
+
+        // undo -> reverts addressbook back to previous state and filtered task list to show all tasks
+        expectedModel.undoTaskManager();
+        assertCommandSuccess(new UndoCommand(), model, commandHistory, UndoCommand.MESSAGE_SUCCESS, expectedModel);
+
+        assertNotEquals(taskToComplete, model.getFilteredTaskList().get(INDEX_FIRST_TASK.getZeroBased()));
+        // redo -> deletes same second task in unfiltered task list
+        expectedModel.redoTaskManager();
+        assertCommandSuccess(new RedoCommand(), model, commandHistory, RedoCommand.MESSAGE_SUCCESS, expectedModel);
+    }
+
```

Comment:
probably add a test for the case when user attempts to complete an already completed task?